packages/sql/src/RequestManager.ts

```diff
@@ -99,28 +108,41 @@ export class RequestManager {
     });
   }
 
-  private _fetch(args: FetchArgs, index: number): Promise<number | undefined> {
-    const [resolve, reject, info, init] = args;
+  private _fetch(requestDefinition: FetchArgs, index: number): Promise<number | undefined> {
+    const {resolve, reject, requestInfo, requestInit, retries_no} = requestDefinition;
 
-    return fetch(info, init)
-      .then((response) => {
+    return fetch(requestInfo, requestInit)
+      .then(async (response) => {
 
         this._retryAfter = this._getRateLimitHeader(response.headers, 'Retry-After', this._retryAfter);
         this._callsLeft = this._getRateLimitHeader(response.headers, 'Carto-Rate-Limit-Remaining', this._callsLeft);
 
-        if (response.status === TOO_MANY_REQUESTS) {
-          // Reschedule
-          this._scheduler();
-          return null;
+        const responseBody = await getResponseBody(response);
+
+        const isTimeoutError = response.status === HTTP_ERRORS.TOO_MANY_REQUESTS &&
+          responseBody.detail === 'datasource';
+
+        if (response.status === HTTP_ERRORS.SERVICE_UNAVAILABLE || isTimeoutError) {
+          requestDefinition.retries_no = retries_no !== NO_RETRY ? retries_no - 1 : MAX_RETRIES;
```

Comment:
In case of `retries_no` is equal to `NO_RETRY`, we are setting `retries_no` to `MAX_RETRIES`. Shouln't it be set to `NO_RETRY` in this case?